modules/loader/src/main/scala/com/snowplowanalytics/snowplow/rdbloader/Main.scala

```diff
@@ -66,21 +66,22 @@ object Main extends IOApp {
 
     Stream.eval_(Manifest.initialize[IO](cli.config.storage, cli.dryRun, env.blocker)) ++
       DataDiscovery.discover[IO](cli.config, env.state)
-        .pauseWhen[IO](env.isBusy)
+        .pauseWhen[IO](Stream.eval[IO, Boolean](env.guard.available.map(_ == 0)))
```

Comment:
I think I see where you're coming from - the `Semaphore` is more clearly states about an intention of 
avoiding a race condition, I agree with that, but I don't think this change works as you think it does. 

`pauseWhen` is all about pausing a _concurrent_ (emphasis one) _stream_ (emphasis two) and for that reason it works with a `Signal` or discrete `Stream` derived from `Signal`. But what you have is `IO[Boolean]` - a single, atomic action, a fact that it is lifted into `Stream` does not add a streaming nature.

To put it another way - `Signal` is an always available (concurrent) effect where a change is propagated immediately and can pause a `Stream` even if no element is pulled (some kind of a backpropagation). `IO` on the other hand is available only once you execute it. As a result, it will pause only after an element has already pulled.

To put it another (a bit blunt) way - `pauseWhen` is designed to work with a `Signal` and you're using it with `IO`.

Reply:
I see, thank you for the explanation.

Would the line below work then? 
```
        .pauseWhen[IO](Stream.awakeEvery[IO](1.second).flatMap(_ => Stream.eval[IO, Boolean](env.guard.available.map(_ == 0))))
```

If I understand correctly, the problem with that approach is that it could take up to 1 second to pause whereas with `SignallingRef` the pause happens as soon as the value inside the `SignallingRef` gets updated?

Reply:
I think it's a bit closer to semantics of `Signal`, but yes - as you mentioned there's always a delay that can be even worse than triggering it. `Signal` is push-based, so you're getting a change as soon as you can, whereas `awakeEvery` is pull-based.

Also, if you ever wanted to implement this, you could use:

```scala
Stream.awakeEvery[IO](1.second).evalMap(_ => env.guard.available.map(_ == 0))
```

Reply:
All clear, great, thank you!